### ISSUE TITLE
Discussion: allow multiple subcommands as argument

### ISSUE THREAD

**Issue by bbottema** [remkop/picocli#434]
(disclaimer, I have very limited experience with / knowledge of this library, my apologies in advance).

You can do this:
```
command --intFlag 1 --stringFlag a --anotherIntFlag 2
```
And this works fine for simple cases. However, what if I have a complex cli structure, where you always need to provide a lot of options.
```
command --param1IntFlag 1 --param1StringFlag a --param1AnotherIntFlag 2 --param2IntFlag 1 --param2StringFlag a --param2AnotherIntFlag 2 --param3IntFlag 1 --param3StringFlag a --param3AnotherIntFlag 2 and so on for 10-20 parameters...
```
Rather verbose. The problem is of course the need for grouping per param, while the multi-values differ in type. Now if I'm understanding correctly, it's not possible to do the following:
```
command --param1 1 a 2 --param2 1 a 2 --param3 1 a 2
```
That's because for multi-value parameters, the value types should be the same and here we are mixing ints and strings. That is _unless I define a custom type for each param that converts the combination of values_. I'm I correct so far?

To work around this, I am actually working with multiple same-level subcommands to achieve this (since with commands, the individual values are converted rather than as a multi-value set):
```
command param1 1 a 2 param2 1 a 2 param3 1 a 2
```
This is possible, because I made every command a sub command of every other sub command, but this produces a rather complex command tree. What would make it a lot easier is if I could instruct picocli to allow for multiple sub commands _on the same level_, so that it would match any command on any level, prioritizing nested matches first (so _A-with-sub-B B_ gets matched over _A-with-sub-C B)_.

Thoughts? Can I solve this a better way already perhaps? I would be very happy to hear this is already possible in a clean way.

**Comment by remkop** [comment]
This reminds me a bit of this: https://github.com/remkop/picocli/issues/358 . Would that meet your requirements?

**Comment by bbottema** [comment]
Well, I'm trying to avoid having to define grouped-argument composites. If picocli would just convert individual values to an expected type, I can just take the array of supplied values and find the appropriate `java.lang.reflect.Method` using reflection.

I would hate to have to update / expand on the composite types each time my API changes. Currently, I have around 10-20 required parameters, with an optional 50 or so more.

**Comment by bbottema** [comment]
Ok well, since I will have so many options of which some of them somewhat complex, is it possible to define usage help for individual options?
```
command --option --help
```
Which would show a help page for the option, where
```
command --help --option
```
would show a help page for the command.

I'm guessing not, in which case for now I'm stuck with the every command is a sub command of every other command.

**Comment by remkop** [comment]
I realized that unix [sendmail](https://linux.die.net/man/8/sendmail.sendmail) has a similar problem: tons and tons of configuration options, and while you want the simple use case to be easy for users, you also want it to be possible for users to do complex things.

sendmail uses `-O option=value` to support a long list of additional options. I guess this is what triggered #435, and perhaps individual option help (`-O:help`) is a good solution.

An alternative is to provide multiple commands, but if you want to build up state in one command to use that state in another command you are entering the realm of an interactive shell-like app. (You may want to look at [jline2](https://github.com/jline/jline2) or jline3(https://github.com/jline/jline3) if that's what you go for eventually.

**Comment by bbottema** [comment]
Actually

> Then you can convert the values that were captured in the list individually in the application.

So this is the problem right there, chicken or the egg: I use reflection to generate the picocli command tree from known Java Methods. When reading it back, I need to find the methods again based on the types I get back from picocli. If I were to tell picocli to define everything as String, then how will I find the right Method back? Number of arguments is not going to be enough!

So this is actually not going to work after all until --options support multi-type multi-values or commands are recognized on all levels. That or I define a custom type handler for each and every multi-type parameter I have (yikes!)

**Comment by bbottema** [comment]
> Perhaps the simplest thing (that is possible now) is to define options that take a range of String parameters

Options still need to be single-value single-type, because I can't show a different helpLabel for each different type in a multi-value list.

So here's how it works now:

```java
.type(String.class)
.helpLabel("emailMessage=FILE replyToAll=BOOL htmlTemplate=STRING")
```

Then later take the String, split up the arguments manually and then convert them. Hairy, but still doable.

**Comment by bbottema** [comment]
Actually too hairy, since splitting up arguments involves more than just breaking on whitespace. Alternatively, multi-value multi-type options would offer a solution but that's not possible (yet). My current work around is a multi-value String collection option, but now I run into the issue with paramLabel. A custom synopsis won't help me completely with this either, since it's still rendered multiple times in the Options: listing. Kind of stuck at the moment...

**Comment by remkop** [comment]
Update: I will try to include a fix for #358 in the upcoming 3.6 release.

**Comment by remkop** [comment]
Picocli 4.0.0-alpha-1 has been released which includes support for repeating composite groups.
See https://picocli.info/#_argument_groups for details. 

I believe this should should meet the requirements of the original use case.

_Please try this and provide feedback. We can still make changes._

_What do you think of the annotations API? What about the programmatic API? Does it work as expected? Are the input validation error messages correct and clear? Is the documentation clear and complete? Anything you want to change or improve? Any other feedback?_
